### PR TITLE
CQL generation

### PR DESCRIPTION
**Pull request by enola-dkfz**
Backward compatibility preserved, both CQL and AST tasks from Lens are handled 
CQL generation logic and CQL code snippets/maps for BBMRI/GBN included
corresponding snippets/maps for [other project using CQL] need to be included in /src/projects/[other project using CQL], the cases added for it in src/projects/mod.rs, feature included in Cargo.toml, and in rust.yml

### CODE REVIEW

File: src/cql.rs [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+                                        format!("@{}", datetime_min.format("%Y-%m-%d"));
+
+                                    condition_string =
+                                        condition_string.replace("{{D1}}", date_str_min.as_str());
+                                    filter_string =
+                                        filter_string.replace("{{D1}}", date_str_min.as_str());
+                                // no condition needed, "" stays ""
+                                } else {
+                                    return Err(FocusError::AstInvalidDateFormat(date_range.min));
+                                }
```
Comment (Threated) [inline src/cql.rs:158]: Just .map_err(...)? after .parse and you dont need the if

File: src/cql.rs [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+                                        format!("@{}", datetime_max.format("%Y-%m-%d"));
+
+                                    condition_string =
+                                        condition_string.replace("{{D2}}", date_str_max.as_str());
+                                    filter_string =
+                                        filter_string.replace("{{D2}}", date_str_max.as_str());
+                                // no condition needed, "" stays ""
+                                } else {
+                                    return Err(FocusError::AstInvalidDateFormat(date_range.max));
+                                }
```
Comment (Threated) [inline src/cql.rs:174]: Same here

File: src/cql.rs [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+                                condition_string = condition_humongous_string + ")";
+
+                                if !filter_string.is_empty() {
+                                    filter_string = filter_humongous_string + ")";
+                                }
+                            }
+                            _ => {
+                                return Err(FocusError::AstOperatorValueMismatch());
+                            }
```
Comment (Threated) [inline src/cql.rs:243]: Maybe have the error say something about what it expected and what it got

File: src/cql.rs [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+                            condition_string = condition_humongous_string + ")";
+
+                            if !filter_string.is_empty() {
+                                filter_string = filter_humongous_string + ")";
+                            }
+                        }
+                        _ => {
+                            return Err(FocusError::AstOperatorValueMismatch());
+                        }
```
Comment (Threated) [inline src/cql.rs:288]: Same here

File: src/cql.rs [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+                    },
+                    ast::ConditionType::NotEquals => { // won't get it from Lens yet
+                    }
+                    ast::ConditionType::Contains => { // won't get it from Lens yet
+                    }
+                    ast::ConditionType::GreaterThan => {} // won't get it from Lens yet
+                    ast::ConditionType::LowerThan => {}   // won't get it from Lens yet
```
Comment (Threated) [inline src/cql.rs:295]: Then I would maybe log something if we get it anyway. FYI you can do `MyEnum::A | MyEnum::B | ... => ...` to match multiple variants to the same code

File: src/cql.rs [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+
+                filter_cond += filter_string.as_str(); // no condition needed, "" can be added with no change
+            } else {
+                return Err(FocusError::AstUnknownCriterion(
+                    condition_key_trans.to_string(),
+                ));
+            }
+            if !filter_cond.is_empty() {
+                //for historical reasons
+            }
```
Comment (Threated) [inline src/cql.rs:312]: Should anything happen here?

File: src/errors.rs [hunk 7]
```diff
@@ -58,7 +68,7 @@ impl FocusError {
         use FocusError::*;
         // TODO: Add more match arms
         match self {
-            DecodeError(_) | ParsingError(_) =>  "Cannot parse query.",
+            DecodeError(_) | SerdeParsingError(_) =>  "Cannot parse query.",
```
Comment (Threated) [inline src/errors.rs:71]: We should still catch the regular Parsing error here

File: src/main.rs [hunk 8]
```diff
@@ -164,39 +167,39 @@ async fn process_task(
...
-                    "Can't run queries with language {} and/or endpoint type {}",
-                    query.lang, CONFIG.endpoint_type
-                ),
-            ))
+        #[derive(Deserialize, Debug)]
+        #[serde(tag = "lang", rename_all = "lowercase")]
+        enum Language {
+            Cql(CqlQuery),
+            Ast(AstQuery)
         }
```
Comment (Threated) [inline src/main.rs:179]: Move it out of the function somewhere where it makes sense

File: src/main.rs [hunk 9]
```diff
@@ -220,6 +223,7 @@ async fn run_cql_query(
     obf_cache: Arc<Mutex<ObfCache>>,
     report_cache: Arc<Mutex<ReportCache>>,
     project: String,
+    generated_from_ast: bool
```
Comment (Threated) [inline src/main.rs:226]: I really dont like this boolean but if you cant come up with a good refactor to not have it feel free to disregard my comment

**Review by Threated — CHANGES_REQUESTED** [review]

File: src/cql.rs (hunk 6 above)
Comment (enola-dkfz) [reply src/cql.rs:312]: at some point it did, then I changed the logic, but in the future after another change maybe there still will, OK, removing it

File: src/main.rs (hunk 9 above)
Comment (enola-dkfz) [reply src/main.rs:226]: It's temporary until all of them are generated by AST, right now we need to keep backward compatibility, so any year now it's gonna be removed

**Pull request merged**